services/shhext/chat/crypto/ethereum_crypto.go

```diff
@@ -0,0 +1,196 @@
+	const (
+		ckInput = 15
+		mkInput = 16
+	)
+
+	h := hmac.New(sha256.New, ck[:])
+
+	_, _ = h.Write([]byte{ckInput})
+	copy(chainKey[:], h.Sum(nil))
+	h.Reset()
+
+	_, _ = h.Write([]byte{mkInput})
+	copy(msgKey[:], h.Sum(nil))
+
+	return chainKey, msgKey
+}
+
+// Encrypt uses a slightly different approach than in the algorithm specification:
+// it uses AES-256-CTR instead of AES-256-CBC for security, ciphertext length and implementation
+// complexity considerations.
+func (c EthereumCrypto) Encrypt(mk dr.Key, plaintext, ad []byte) []byte {
+	encKey, authKey, iv := c.deriveEncKeys(mk)
+
+	ciphertext := make([]byte, aes.BlockSize+len(plaintext))
+	copy(ciphertext, iv[:])
+
+	var (
+		block, _ = aes.NewCipher(encKey[:]) // No error will occur here as encKey is guaranteed to be 32 bytes.
+		stream   = cipher.NewCTR(block, iv[:])
+	)
+	stream.XORKeyStream(ciphertext[aes.BlockSize:], plaintext)
+
+	return append(ciphertext, c.computeSignature(authKey[:], ciphertext, ad)...)
+}
+
+// See the Crypto interface.
+func (c EthereumCrypto) Decrypt(mk dr.Key, authCiphertext, ad []byte) ([]byte, error) {
+	var (
+		l          = len(authCiphertext)
+		ciphertext = authCiphertext[:l-sha256.Size]
+		signature  = authCiphertext[l-sha256.Size:]
+	)
+
+	// Check the signature.
+	encKey, authKey, _ := c.deriveEncKeys(mk)
+
+	if s := c.computeSignature(authKey[:], ciphertext, ad); !bytes.Equal(s, signature) {
+		return nil, fmt.Errorf("invalid signature")
+	}
+
+	// Decrypt.
+	var (
+		block, _  = aes.NewCipher(encKey[:]) // No error will occur here as encKey is guaranteed to be 32 bytes.
+		stream    = cipher.NewCTR(block, ciphertext[:aes.BlockSize])
+		plaintext = make([]byte, len(ciphertext[aes.BlockSize:]))
+	)
+	stream.XORKeyStream(plaintext, ciphertext[aes.BlockSize:])
+
+	return plaintext, nil
+}
+
+// deriveEncKeys derive keys for message encryption and decryption. Returns (encKey, authKey, iv, err).
+func (c EthereumCrypto) deriveEncKeys(mk dr.Key) (encKey dr.Key, authKey dr.Key, iv [16]byte) {
+	// First, derive encryption and authentication key out of mk.
+	salt := make([]byte, 32)
+	var (
+		r   = hkdf.New(sha256.New, mk[:], salt, []byte("pcwSByyx2CRdryCffXJwy7xgVZWtW5Sh"))
+		buf = make([]byte, 80)
+	)
+
+	// The only error here is an entropy limit which won't be reached for such a short buffer.
+	_, _ = io.ReadFull(r, buf)
+
+	copy(encKey[:], buf[0:32])
+	copy(authKey[:], buf[32:64])
+	copy(iv[:], buf[64:80])
+	return
+}
+
+func (c EthereumCrypto) computeSignature(authKey, ciphertext, associatedData []byte) []byte {
+	h := hmac.New(sha256.New, authKey)
+	_, _ = h.Write(associatedData)
+	_, _ = h.Write(ciphertext)
+	return h.Sum(nil)
+}
+
+type DHPair struct {
+	PrvKey dr.Key
+	PubKey dr.Key
+}
+
+func (p DHPair) PrivateKey() dr.Key {
+	return p.PrvKey
+}
+
+func (p DHPair) PublicKey() dr.Key {
+	return p.PubKey
+}
+
+func (p DHPair) String() string {
+	return fmt.Sprintf("{privateKey: %s publicKey: %s}", p.PrvKey, p.PubKey)
```

Comment:
You sure we want to log private key ever? It's a bit dangerous as people may use `String()` for debug purposes or in loggers. This way a private key can end up in the log files.